src/main/java/org/jenkinsci/plugins/workflow/cps/EnvActionImpl.java

```diff
@@ -113,19 +114,23 @@ private EnvActionImpl() {
         owner = r;
     }
 
+    public static @Nonnull EnvActionImpl forRun(@Nonnull Run<?,?> run) {
+        EnvActionImpl action = run.getAction(EnvActionImpl.class);
+        if (action == null) {
+            action = new EnvActionImpl();
+            run.addAction(action);
```

Comment:
Suppose it could `run.save()`. The build is likely to get saved for some other reason soon afterwards anyway.

I would note that numerous calls to `Run.addAction` in Jenkins core decline to call `save()`, with no apparent ill effect. Probably this override could call `save()` on itself.